// rookout/processor/namespaces/InspectorTracebackNamespace.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InspectorTracebackNamespace = void 0;

var _Namespace = _interopRequireDefault(require("./Namespace"));

var _InspectorFrameNamespace = _interopRequireDefault(require("./InspectorFrameNamespace"));

var _JSObjectNamespace = _interopRequireDefault(require("./JSObjectNamespace"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const variant_pb = require('../../protobuf/variant_pb');

const code_object_lineno_field_number = 4;
const code_object_name_index_in_cache_field_number = 5;
const code_object_module_index_in_cache_field_number = 6;
const code_object_filename_index_in_cache_field_number = 7;

class InspectorTracebackNamespace extends _Namespace.default {
  constructor(inspector, callFrames, depth) {
    super();
    this.inspector = inspector;
    this.callFrames = callFrames;
    this.depth = depth;
  }

  readKey(key) {
    return new _InspectorFrameNamespace.default(this.inspector, this.callFrames[key]);
  }

  callMethod(name, args) {
    switch (name) {
      case "size":
        return new _JSObjectNamespace.default(this.depth);

      default:
        return super.callMethod(name, args);
    }
  }

  dump(variant, getStringIndexInCacheCallable = null) {
    variant.setVariantType(variant_pb.Variant.Type.VARIANT_TRACEBACK);
    let traceback = new variant_pb.Variant.Traceback();

    for (let i = 0; i < this.depth; ++i) {
      const position = this.inspector.resolvePosition(this.callFrames[i]);
      let frame = new variant_pb.Variant.CodeObject();

      if (getStringIndexInCacheCallable === null) {
        frame.setName(position.function);
        frame.setFilename(position.filename);
        frame.setModule(position.filename);
      } else {
        frame.setNameIndexInCache(getStringIndexInCacheCallable(position.function || '<function missing>'));
        frame.setFilenameIndexInCache(getStringIndexInCacheCallable(position.filename || '<file missing>'));
        frame.setModuleIndexInCache(getStringIndexInCacheCallable(position.filename || '<module missing>'));
      }

      frame.setLineno(position.line);
      traceback.addLocations(frame);
    }

    variant.setTraceback(traceback);
  }

  streamDump(variant2Writer, getStringIndexInCacheCallable) {
    variant2Writer.writeVariantType(variant_pb.Variant.Type.VARIANT_TRACEBACK);
    let positions = [];

    for (let i = 0; i < this.depth; ++i) {
      positions.push(this.inspector.resolvePosition(this.callFrames[i]));
    }

    variant2Writer.writeCodeValues(positions, (obj, writer) => {
      writer.writeUint32(code_object_lineno_field_number, obj.line);
      writer.writeUint32(code_object_name_index_in_cache_field_number, getStringIndexInCacheCallable(obj.function || '<function missing>'));
      writer.writeUint32(code_object_filename_index_in_cache_field_number, getStringIndexInCacheCallable(obj.filename || '<file missing>'));
      writer.writeUint32(code_object_module_index_in_cache_field_number, getStringIndexInCacheCallable(obj.filename || '<module missing>'));
    });
  }

}

exports.InspectorTracebackNamespace = InspectorTracebackNamespace;
