// rookout/processor/paths/ArithmeticPathInternal.js
"use strict";

/**
 * This file was generated from canopy\ArithmeticPath.peg
 * See https://canopy.jcoglan.com/ for documentation
 */
(function () {
  'use strict';

  function TreeNode(text, offset, elements) {
    this.text = text;
    this.offset = offset;
    this.elements = elements;
  }

  TreeNode.prototype.forEach = function (block, context) {
    for (var el = this.elements, i = 0, n = el.length; i < n; i++) {
      block.call(context, el[i], i, el);
    }
  };

  if (typeof Symbol !== 'undefined' && Symbol.iterator) {
    TreeNode.prototype[Symbol.iterator] = function () {
      return this.elements[Symbol.iterator]();
    };
  }

  var TreeNode1 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['atom'] = elements[0];
  };

  inherit(TreeNode1, TreeNode);

  var TreeNode2 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['opt_'] = elements[0];
    this['atom'] = elements[1];
  };

  inherit(TreeNode2, TreeNode);

  var TreeNode3 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['unicode_set'] = elements[0];
  };

  inherit(TreeNode3, TreeNode);

  var TreeNode4 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['variable'] = elements[1];
  };

  inherit(TreeNode4, TreeNode);

  var TreeNode5 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['unicode_set'] = elements[0];
  };

  inherit(TreeNode5, TreeNode);

  var TreeNode6 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['unicode_set'] = elements[1];
  };

  inherit(TreeNode6, TreeNode);

  var TreeNode7 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[3];
  };

  inherit(TreeNode7, TreeNode);

  var TreeNode8 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[4];
    this['comp_expression'] = elements[2];
  };

  inherit(TreeNode8, TreeNode);

  var TreeNode9 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[4];
  };

  inherit(TreeNode9, TreeNode);

  var TreeNode10 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[4];
  };

  inherit(TreeNode10, TreeNode);

  var TreeNode11 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[6];
  };

  inherit(TreeNode11, TreeNode);

  var TreeNode12 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['atom'] = elements[0];
    this['__'] = elements[1];
  };

  inherit(TreeNode12, TreeNode);

  var TreeNode13 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[3];
    this['atom'] = elements[2];
  };

  inherit(TreeNode13, TreeNode);

  var TreeNode14 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[4];
  };

  inherit(TreeNode14, TreeNode);

  var TreeNode15 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[3];
  };

  inherit(TreeNode15, TreeNode);

  var TreeNode16 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[5];
  };

  inherit(TreeNode16, TreeNode);

  var TreeNode17 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[2];
    this['data'] = elements[1];
  };

  inherit(TreeNode17, TreeNode);

  var TreeNode18 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[2];
    this['data'] = elements[1];
  };

  inherit(TreeNode18, TreeNode);

  var TreeNode19 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[2];
    this['data'] = elements[1];
  };

  inherit(TreeNode19, TreeNode);

  var TreeNode20 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[2];
    this['data'] = elements[1];
  };

  inherit(TreeNode20, TreeNode);

  var TreeNode21 = function (text, offset, elements) {
    TreeNode.apply(this, arguments);
    this['__'] = elements[2];
    this['data'] = elements[1];
  };

  inherit(TreeNode21, TreeNode);
  var FAILURE = {};
  var Grammar = {
    _read_comp_expression() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._comp_expression = this._cache._comp_expression || {};
      var cached = this._cache._comp_expression[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(2);
      var address1 = FAILURE;
      address1 = this._read_atom();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset,
            elements1 = [],
            address3 = null;

        while (true) {
          var index3 = this._offset,
              elements2 = new Array(2);
          var address4 = FAILURE;
          address4 = this._read_opt_();

          if (address4 !== FAILURE) {
            elements2[0] = address4;
            var address5 = FAILURE;
            address5 = this._read_atom();

            if (address5 !== FAILURE) {
              elements2[1] = address5;
            } else {
              elements2 = null;
              this._offset = index3;
            }
          } else {
            elements2 = null;
            this._offset = index3;
          }

          if (elements2 === null) {
            address3 = FAILURE;
          } else {
            address3 = new TreeNode2(this._input.substring(index3, this._offset), index3, elements2);
            this._offset = this._offset;
          }

          if (address3 !== FAILURE) {
            elements1.push(address3);
          } else {
            break;
          }
        }

        if (elements1.length >= 0) {
          address2 = new TreeNode(this._input.substring(index2, this._offset), index2, elements1);
          this._offset = this._offset;
        } else {
          address2 = FAILURE;
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_comp_exp(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._comp_expression[index0] = [address0, this._offset];
      return address0;
    },

    _read_atom() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._atom = this._cache._atom || {};
      var cached = this._cache._atom[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset;
      address0 = this._read_float_();

      if (address0 === FAILURE) {
        this._offset = index1;
        address0 = this._read_number();

        if (address0 === FAILURE) {
          this._offset = index1;
          address0 = this._read_string();

          if (address0 === FAILURE) {
            this._offset = index1;
            address0 = this._read_apostrophe_string();

            if (address0 === FAILURE) {
              this._offset = index1;
              address0 = this._read_list();

              if (address0 === FAILURE) {
                this._offset = index1;
                address0 = this._read_chr();

                if (address0 === FAILURE) {
                  this._offset = index1;
                  address0 = this._read_boolean_();

                  if (address0 === FAILURE) {
                    this._offset = index1;
                    address0 = this._read_null_();

                    if (address0 === FAILURE) {
                      this._offset = index1;
                      address0 = this._read_nan_();

                      if (address0 === FAILURE) {
                        this._offset = index1;
                        address0 = this._read_undefined_();

                        if (address0 === FAILURE) {
                          this._offset = index1;
                          address0 = this._read_comp_expression_ex();

                          if (address0 === FAILURE) {
                            this._offset = index1;
                            address0 = this._read_namespace();

                            if (address0 === FAILURE) {
                              this._offset = index1;
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }

      this._cache._atom[index0] = [address0, this._offset];
      return address0;
    },

    _read_unicode_set() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._unicode_set = this._cache._unicode_set || {};
      var cached = this._cache._unicode_set[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var chunk0 = null,
          max0 = this._offset + 1;

      if (max0 <= this._inputSize) {
        chunk0 = this._input.substring(this._offset, max0);
      }

      if (chunk0 !== null && /^[a-zA-Z_$@]/.test(chunk0)) {
        address0 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
        this._offset = this._offset + 1;
      } else {
        address0 = FAILURE;

        if (this._offset > this._failure) {
          this._failure = this._offset;
          this._expected = [];
        }

        if (this._offset === this._failure) {
          this._expected.push(['Maps::unicode_set', '[a-zA-Z_$@]']);
        }
      }

      this._cache._unicode_set[index0] = [address0, this._offset];
      return address0;
    },

    _read_unicode_set_with_numbers() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._unicode_set_with_numbers = this._cache._unicode_set_with_numbers || {};
      var cached = this._cache._unicode_set_with_numbers[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var chunk0 = null,
          max0 = this._offset + 1;

      if (max0 <= this._inputSize) {
        chunk0 = this._input.substring(this._offset, max0);
      }

      if (chunk0 !== null && /^[a-zA-Z0-9_$@]/.test(chunk0)) {
        address0 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
        this._offset = this._offset + 1;
      } else {
        address0 = FAILURE;

        if (this._offset > this._failure) {
          this._failure = this._offset;
          this._expected = [];
        }

        if (this._offset === this._failure) {
          this._expected.push(['Maps::unicode_set_with_numbers', '[a-zA-Z0-9_$@]']);
        }
      }

      this._cache._unicode_set_with_numbers[index0] = [address0, this._offset];
      return address0;
    },

    _read_variable() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._variable = this._cache._variable || {};
      var cached = this._cache._variable[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(2);
      var address1 = FAILURE;
      address1 = this._read_unicode_set();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset,
            elements1 = [],
            address3 = null;

        while (true) {
          address3 = this._read_unicode_set_with_numbers();

          if (address3 !== FAILURE) {
            elements1.push(address3);
          } else {
            break;
          }
        }

        if (elements1.length >= 0) {
          address2 = new TreeNode(this._input.substring(index2, this._offset), index2, elements1);
          this._offset = this._offset;
        } else {
          address2 = FAILURE;
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_attribute(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._variable[index0] = [address0, this._offset];
      return address0;
    },

    _read_variable_access() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._variable_access = this._cache._variable_access || {};
      var cached = this._cache._variable_access[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(2);
      var address1 = FAILURE;
      var chunk0 = null,
          max0 = this._offset + 1;

      if (max0 <= this._inputSize) {
        chunk0 = this._input.substring(this._offset, max0);
      }

      if (chunk0 === '.') {
        address1 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
        this._offset = this._offset + 1;
      } else {
        address1 = FAILURE;

        if (this._offset > this._failure) {
          this._failure = this._offset;
          this._expected = [];
        }

        if (this._offset === this._failure) {
          this._expected.push(['Maps::variable_access', '"."']);
        }
      }

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        address2 = this._read_variable();

        if (address2 !== FAILURE) {
          elements0[1] = address2;
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_attribute_operation(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._variable_access[index0] = [address0, this._offset];
      return address0;
    },

    _read_map_access() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._map_access = this._cache._map_access || {};
      var cached = this._cache._map_access[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(3);
      var address1 = FAILURE;
      var chunk0 = null,
          max0 = this._offset + 1;

      if (max0 <= this._inputSize) {
        chunk0 = this._input.substring(this._offset, max0);
      }

      if (chunk0 === '[') {
        address1 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
        this._offset = this._offset + 1;
      } else {
        address1 = FAILURE;

        if (this._offset > this._failure) {
          this._failure = this._offset;
          this._expected = [];
        }

        if (this._offset === this._failure) {
          this._expected.push(['Maps::map_access', '"["']);
        }
      }

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        address2 = this._read_chr();

        if (address2 === FAILURE) {
          this._offset = index2;
          address2 = this._read_string();

          if (address2 === FAILURE) {
            this._offset = index2;
            address2 = this._read_apostrophe_string();

            if (address2 === FAILURE) {
              this._offset = index2;
              address2 = this._read_number();

              if (address2 === FAILURE) {
                this._offset = index2;
              }
            }
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var chunk1 = null,
              max1 = this._offset + 1;

          if (max1 <= this._inputSize) {
            chunk1 = this._input.substring(this._offset, max1);
          }

          if (chunk1 === ']') {
            address3 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
            this._offset = this._offset + 1;
          } else {
            address3 = FAILURE;

            if (this._offset > this._failure) {
              this._failure = this._offset;
              this._expected = [];
            }

            if (this._offset === this._failure) {
              this._expected.push(['Maps::map_access', '"]"']);
            }
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_lookup_operation(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._map_access[index0] = [address0, this._offset];
      return address0;
    },

    _read_func_call() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._func_call = this._cache._func_call || {};
      var cached = this._cache._func_call[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(5);
      var address1 = FAILURE;
      address1 = this._read_unicode_set();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset,
            elements1 = [],
            address3 = null;

        while (true) {
          address3 = this._read_unicode_set_with_numbers();

          if (address3 !== FAILURE) {
            elements1.push(address3);
          } else {
            break;
          }
        }

        if (elements1.length >= 0) {
          address2 = new TreeNode(this._input.substring(index2, this._offset), index2, elements1);
          this._offset = this._offset;
        } else {
          address2 = FAILURE;
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address4 = FAILURE;
          var chunk0 = null,
              max0 = this._offset + 1;

          if (max0 <= this._inputSize) {
            chunk0 = this._input.substring(this._offset, max0);
          }

          if (chunk0 === '(') {
            address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
            this._offset = this._offset + 1;
          } else {
            address4 = FAILURE;

            if (this._offset > this._failure) {
              this._failure = this._offset;
              this._expected = [];
            }

            if (this._offset === this._failure) {
              this._expected.push(['Maps::func_call', '"("']);
            }
          }

          if (address4 !== FAILURE) {
            elements0[2] = address4;
            var address5 = FAILURE;
            var index3 = this._offset;
            address5 = this._read_atom();

            if (address5 === FAILURE) {
              address5 = new TreeNode(this._input.substring(index3, index3), index3, []);
              this._offset = index3;
            }

            if (address5 !== FAILURE) {
              elements0[3] = address5;
              var address6 = FAILURE;
              var chunk1 = null,
                  max1 = this._offset + 1;

              if (max1 <= this._inputSize) {
                chunk1 = this._input.substring(this._offset, max1);
              }

              if (chunk1 === ')') {
                address6 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                this._offset = this._offset + 1;
              } else {
                address6 = FAILURE;

                if (this._offset > this._failure) {
                  this._failure = this._offset;
                  this._expected = [];
                }

                if (this._offset === this._failure) {
                  this._expected.push(['Maps::func_call', '")"']);
                }
              }

              if (address6 !== FAILURE) {
                elements0[4] = address6;
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_function_operation(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._func_call[index0] = [address0, this._offset];
      return address0;
    },

    _read_func_call_access() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._func_call_access = this._cache._func_call_access || {};
      var cached = this._cache._func_call_access[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(6);
      var address1 = FAILURE;
      var chunk0 = null,
          max0 = this._offset + 1;

      if (max0 <= this._inputSize) {
        chunk0 = this._input.substring(this._offset, max0);
      }

      if (chunk0 === '.') {
        address1 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
        this._offset = this._offset + 1;
      } else {
        address1 = FAILURE;

        if (this._offset > this._failure) {
          this._failure = this._offset;
          this._expected = [];
        }

        if (this._offset === this._failure) {
          this._expected.push(['Maps::func_call_access', '"."']);
        }
      }

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        address2 = this._read_unicode_set();

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var index2 = this._offset,
              elements1 = [],
              address4 = null;

          while (true) {
            address4 = this._read_unicode_set_with_numbers();

            if (address4 !== FAILURE) {
              elements1.push(address4);
            } else {
              break;
            }
          }

          if (elements1.length >= 0) {
            address3 = new TreeNode(this._input.substring(index2, this._offset), index2, elements1);
            this._offset = this._offset;
          } else {
            address3 = FAILURE;
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address5 = FAILURE;
            var chunk1 = null,
                max1 = this._offset + 1;

            if (max1 <= this._inputSize) {
              chunk1 = this._input.substring(this._offset, max1);
            }

            if (chunk1 === '(') {
              address5 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address5 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::func_call_access', '"("']);
              }
            }

            if (address5 !== FAILURE) {
              elements0[3] = address5;
              var address6 = FAILURE;
              var index3 = this._offset;
              address6 = this._read_atom();

              if (address6 === FAILURE) {
                address6 = new TreeNode(this._input.substring(index3, index3), index3, []);
                this._offset = index3;
              }

              if (address6 !== FAILURE) {
                elements0[4] = address6;
                var address7 = FAILURE;
                var chunk2 = null,
                    max2 = this._offset + 1;

                if (max2 <= this._inputSize) {
                  chunk2 = this._input.substring(this._offset, max2);
                }

                if (chunk2 === ')') {
                  address7 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                  this._offset = this._offset + 1;
                } else {
                  address7 = FAILURE;

                  if (this._offset > this._failure) {
                    this._failure = this._offset;
                    this._expected = [];
                  }

                  if (this._offset === this._failure) {
                    this._expected.push(['Maps::func_call_access', '")"']);
                  }
                }

                if (address7 !== FAILURE) {
                  elements0[5] = address7;
                } else {
                  elements0 = null;
                  this._offset = index1;
                }
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_function_operation_access(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._func_call_access[index0] = [address0, this._offset];
      return address0;
    },

    _read_namespace() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._namespace = this._cache._namespace || {};
      var cached = this._cache._namespace[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(4);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        address2 = this._read_func_call();

        if (address2 === FAILURE) {
          this._offset = index2;
          address2 = this._read_variable();

          if (address2 === FAILURE) {
            this._offset = index2;
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var index3 = this._offset,
              elements1 = [],
              address4 = null;

          while (true) {
            var index4 = this._offset;
            address4 = this._read_func_call_access();

            if (address4 === FAILURE) {
              this._offset = index4;
              address4 = this._read_map_access();

              if (address4 === FAILURE) {
                this._offset = index4;
                address4 = this._read_variable_access();

                if (address4 === FAILURE) {
                  this._offset = index4;
                }
              }
            }

            if (address4 !== FAILURE) {
              elements1.push(address4);
            } else {
              break;
            }
          }

          if (elements1.length >= 0) {
            address3 = new TreeNode(this._input.substring(index3, this._offset), index3, elements1);
            this._offset = this._offset;
          } else {
            address3 = FAILURE;
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address5 = FAILURE;
            address5 = this._read___();

            if (address5 !== FAILURE) {
              elements0[3] = address5;
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_and_execute_namespace_operation(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._namespace[index0] = [address0, this._offset];
      return address0;
    },

    _read_comp_expression_ex() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._comp_expression_ex = this._cache._comp_expression_ex || {};
      var cached = this._cache._comp_expression_ex[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(5);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === '(') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::comp_expression_ex', '"("']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read_comp_expression();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address4 = FAILURE;
            var chunk1 = null,
                max1 = this._offset + 1;

            if (max1 <= this._inputSize) {
              chunk1 = this._input.substring(this._offset, max1);
            }

            if (chunk1 === ')') {
              address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address4 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::comp_expression_ex', '")"']);
              }
            }

            if (address4 !== FAILURE) {
              elements0[3] = address4;
              var address5 = FAILURE;
              address5 = this._read___();

              if (address5 !== FAILURE) {
                elements0[4] = address5;
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_comp_exp_ex(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._comp_expression_ex[index0] = [address0, this._offset];
      return address0;
    },

    _read_string() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._string = this._cache._string || {};
      var cached = this._cache._string[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(5);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 !== null && /^[\"]/.test(chunk0)) {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::string', '[\\"]']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var index2 = this._offset,
              elements1 = [],
              address4 = null;

          while (true) {
            var chunk1 = null,
                max1 = this._offset + 1;

            if (max1 <= this._inputSize) {
              chunk1 = this._input.substring(this._offset, max1);
            }

            if (chunk1 !== null && /^[ !#-~]/.test(chunk1)) {
              address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address4 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::string', '[ !#-~]']);
              }
            }

            if (address4 !== FAILURE) {
              elements1.push(address4);
            } else {
              break;
            }
          }

          if (elements1.length >= 0) {
            address3 = new TreeNode(this._input.substring(index2, this._offset), index2, elements1);
            this._offset = this._offset;
          } else {
            address3 = FAILURE;
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address5 = FAILURE;
            var chunk2 = null,
                max2 = this._offset + 1;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 !== null && /^[\"]/.test(chunk2)) {
              address5 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address5 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::string', '[\\"]']);
              }
            }

            if (address5 !== FAILURE) {
              elements0[3] = address5;
              var address6 = FAILURE;
              address6 = this._read___();

              if (address6 !== FAILURE) {
                elements0[4] = address6;
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_string(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._string[index0] = [address0, this._offset];
      return address0;
    },

    _read_apostrophe_string() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._apostrophe_string = this._cache._apostrophe_string || {};
      var cached = this._cache._apostrophe_string[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(5);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === '\'') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::apostrophe_string', '"\'"']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var index2 = this._offset,
              elements1 = [],
              address4 = null;

          while (true) {
            var chunk1 = null,
                max1 = this._offset + 1;

            if (max1 <= this._inputSize) {
              chunk1 = this._input.substring(this._offset, max1);
            }

            if (chunk1 !== null && /^[^']/.test(chunk1)) {
              address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address4 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::apostrophe_string', '[^\']']);
              }
            }

            if (address4 !== FAILURE) {
              elements1.push(address4);
            } else {
              break;
            }
          }

          if (elements1.length >= 0) {
            address3 = new TreeNode(this._input.substring(index2, this._offset), index2, elements1);
            this._offset = this._offset;
          } else {
            address3 = FAILURE;
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address5 = FAILURE;
            var chunk2 = null,
                max2 = this._offset + 1;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 === '\'') {
              address5 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address5 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::apostrophe_string', '"\'"']);
              }
            }

            if (address5 !== FAILURE) {
              elements0[3] = address5;
              var address6 = FAILURE;
              address6 = this._read___();

              if (address6 !== FAILURE) {
                elements0[4] = address6;
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_apostrophe_string(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._apostrophe_string[index0] = [address0, this._offset];
      return address0;
    },

    _read_list() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._list = this._cache._list || {};
      var cached = this._cache._list[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(7);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === '[') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::list', '"["']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read___();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address4 = FAILURE;
            var index2 = this._offset;
            var index3 = this._offset,
                elements1 = new Array(3);
            var address5 = FAILURE;
            address5 = this._read_atom();

            if (address5 !== FAILURE) {
              elements1[0] = address5;
              var address6 = FAILURE;
              address6 = this._read___();

              if (address6 !== FAILURE) {
                elements1[1] = address6;
                var address7 = FAILURE;
                var index4 = this._offset,
                    elements2 = [],
                    address8 = null;

                while (true) {
                  var index5 = this._offset,
                      elements3 = new Array(4);
                  var address9 = FAILURE;
                  var chunk1 = null,
                      max1 = this._offset + 1;

                  if (max1 <= this._inputSize) {
                    chunk1 = this._input.substring(this._offset, max1);
                  }

                  if (chunk1 === ',') {
                    address9 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                    this._offset = this._offset + 1;
                  } else {
                    address9 = FAILURE;

                    if (this._offset > this._failure) {
                      this._failure = this._offset;
                      this._expected = [];
                    }

                    if (this._offset === this._failure) {
                      this._expected.push(['Maps::list', '","']);
                    }
                  }

                  if (address9 !== FAILURE) {
                    elements3[0] = address9;
                    var address10 = FAILURE;
                    address10 = this._read___();

                    if (address10 !== FAILURE) {
                      elements3[1] = address10;
                      var address11 = FAILURE;
                      address11 = this._read_atom();

                      if (address11 !== FAILURE) {
                        elements3[2] = address11;
                        var address12 = FAILURE;
                        address12 = this._read___();

                        if (address12 !== FAILURE) {
                          elements3[3] = address12;
                        } else {
                          elements3 = null;
                          this._offset = index5;
                        }
                      } else {
                        elements3 = null;
                        this._offset = index5;
                      }
                    } else {
                      elements3 = null;
                      this._offset = index5;
                    }
                  } else {
                    elements3 = null;
                    this._offset = index5;
                  }

                  if (elements3 === null) {
                    address8 = FAILURE;
                  } else {
                    address8 = new TreeNode13(this._input.substring(index5, this._offset), index5, elements3);
                    this._offset = this._offset;
                  }

                  if (address8 !== FAILURE) {
                    elements2.push(address8);
                  } else {
                    break;
                  }
                }

                if (elements2.length >= 0) {
                  address7 = new TreeNode(this._input.substring(index4, this._offset), index4, elements2);
                  this._offset = this._offset;
                } else {
                  address7 = FAILURE;
                }

                if (address7 !== FAILURE) {
                  elements1[2] = address7;
                } else {
                  elements1 = null;
                  this._offset = index3;
                }
              } else {
                elements1 = null;
                this._offset = index3;
              }
            } else {
              elements1 = null;
              this._offset = index3;
            }

            if (elements1 === null) {
              address4 = FAILURE;
            } else {
              address4 = new TreeNode12(this._input.substring(index3, this._offset), index3, elements1);
              this._offset = this._offset;
            }

            if (address4 === FAILURE) {
              address4 = new TreeNode(this._input.substring(index2, index2), index2, []);
              this._offset = index2;
            }

            if (address4 !== FAILURE) {
              elements0[3] = address4;
              var address13 = FAILURE;
              address13 = this._read___();

              if (address13 !== FAILURE) {
                elements0[4] = address13;
                var address14 = FAILURE;
                var chunk2 = null,
                    max2 = this._offset + 1;

                if (max2 <= this._inputSize) {
                  chunk2 = this._input.substring(this._offset, max2);
                }

                if (chunk2 === ']') {
                  address14 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                  this._offset = this._offset + 1;
                } else {
                  address14 = FAILURE;

                  if (this._offset > this._failure) {
                    this._failure = this._offset;
                    this._expected = [];
                  }

                  if (this._offset === this._failure) {
                    this._expected.push(['Maps::list', '"]"']);
                  }
                }

                if (address14 !== FAILURE) {
                  elements0[5] = address14;
                  var address15 = FAILURE;
                  address15 = this._read___();

                  if (address15 !== FAILURE) {
                    elements0[6] = address15;
                  } else {
                    elements0 = null;
                    this._offset = index1;
                  }
                } else {
                  elements0 = null;
                  this._offset = index1;
                }
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_list(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._list[index0] = [address0, this._offset];
      return address0;
    },

    _read_chr() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._chr = this._cache._chr || {};
      var cached = this._cache._chr[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(5);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 !== null && /^[']/.test(chunk0)) {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::chr', '[\']']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var chunk1 = null,
              max1 = this._offset + 1;

          if (max1 <= this._inputSize) {
            chunk1 = this._input.substring(this._offset, max1);
          }

          if (chunk1 !== null && /^[ !#-~]/.test(chunk1)) {
            address3 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
            this._offset = this._offset + 1;
          } else {
            address3 = FAILURE;

            if (this._offset > this._failure) {
              this._failure = this._offset;
              this._expected = [];
            }

            if (this._offset === this._failure) {
              this._expected.push(['Maps::chr', '[ !#-~]']);
            }
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address4 = FAILURE;
            var chunk2 = null,
                max2 = this._offset + 1;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 !== null && /^[']/.test(chunk2)) {
              address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address4 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::chr', '[\']']);
              }
            }

            if (address4 !== FAILURE) {
              elements0[3] = address4;
              var address5 = FAILURE;
              address5 = this._read___();

              if (address5 !== FAILURE) {
                elements0[4] = address5;
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_char(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._chr[index0] = [address0, this._offset];
      return address0;
    },

    _read_number() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._number = this._cache._number || {};
      var cached = this._cache._number[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(4);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === '-') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::number', '"-"']);
          }
        }

        if (address2 === FAILURE) {
          address2 = new TreeNode(this._input.substring(index2, index2), index2, []);
          this._offset = index2;
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var index3 = this._offset,
              elements1 = [],
              address4 = null;

          while (true) {
            var chunk1 = null,
                max1 = this._offset + 1;

            if (max1 <= this._inputSize) {
              chunk1 = this._input.substring(this._offset, max1);
            }

            if (chunk1 !== null && /^[0-9]/.test(chunk1)) {
              address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address4 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::number', '[0-9]']);
              }
            }

            if (address4 !== FAILURE) {
              elements1.push(address4);
            } else {
              break;
            }
          }

          if (elements1.length >= 1) {
            address3 = new TreeNode(this._input.substring(index3, this._offset), index3, elements1);
            this._offset = this._offset;
          } else {
            address3 = FAILURE;
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address5 = FAILURE;
            address5 = this._read___();

            if (address5 !== FAILURE) {
              elements0[3] = address5;
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_number(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._number[index0] = [address0, this._offset];
      return address0;
    },

    _read_float_() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._float_ = this._cache._float_ || {};
      var cached = this._cache._float_[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(6);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === '-') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::float_', '"-"']);
          }
        }

        if (address2 === FAILURE) {
          address2 = new TreeNode(this._input.substring(index2, index2), index2, []);
          this._offset = index2;
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          var index3 = this._offset,
              elements1 = [],
              address4 = null;

          while (true) {
            var chunk1 = null,
                max1 = this._offset + 1;

            if (max1 <= this._inputSize) {
              chunk1 = this._input.substring(this._offset, max1);
            }

            if (chunk1 !== null && /^[0-9]/.test(chunk1)) {
              address4 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address4 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::float_', '[0-9]']);
              }
            }

            if (address4 !== FAILURE) {
              elements1.push(address4);
            } else {
              break;
            }
          }

          if (elements1.length >= 1) {
            address3 = new TreeNode(this._input.substring(index3, this._offset), index3, elements1);
            this._offset = this._offset;
          } else {
            address3 = FAILURE;
          }

          if (address3 !== FAILURE) {
            elements0[2] = address3;
            var address5 = FAILURE;
            var chunk2 = null,
                max2 = this._offset + 1;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 === '.') {
              address5 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address5 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::float_', '"."']);
              }
            }

            if (address5 !== FAILURE) {
              elements0[3] = address5;
              var address6 = FAILURE;
              var index4 = this._offset,
                  elements2 = [],
                  address7 = null;

              while (true) {
                var chunk3 = null,
                    max3 = this._offset + 1;

                if (max3 <= this._inputSize) {
                  chunk3 = this._input.substring(this._offset, max3);
                }

                if (chunk3 !== null && /^[0-9]/.test(chunk3)) {
                  address7 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                  this._offset = this._offset + 1;
                } else {
                  address7 = FAILURE;

                  if (this._offset > this._failure) {
                    this._failure = this._offset;
                    this._expected = [];
                  }

                  if (this._offset === this._failure) {
                    this._expected.push(['Maps::float_', '[0-9]']);
                  }
                }

                if (address7 !== FAILURE) {
                  elements2.push(address7);
                } else {
                  break;
                }
              }

              if (elements2.length >= 1) {
                address6 = new TreeNode(this._input.substring(index4, this._offset), index4, elements2);
                this._offset = this._offset;
              } else {
                address6 = FAILURE;
              }

              if (address6 !== FAILURE) {
                elements0[4] = address6;
                var address8 = FAILURE;
                address8 = this._read___();

                if (address8 !== FAILURE) {
                  elements0[5] = address8;
                } else {
                  elements0 = null;
                  this._offset = index1;
                }
              } else {
                elements0 = null;
                this._offset = index1;
              }
            } else {
              elements0 = null;
              this._offset = index1;
            }
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_float(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._float_[index0] = [address0, this._offset];
      return address0;
    },

    _read_boolean_() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._boolean_ = this._cache._boolean_ || {};
      var cached = this._cache._boolean_[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(3);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        var chunk0 = null,
            max0 = this._offset + 4;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === 'True') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 4), this._offset, []);
          this._offset = this._offset + 4;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::boolean_', '"True"']);
          }
        }

        if (address2 === FAILURE) {
          this._offset = index2;
          var chunk1 = null,
              max1 = this._offset + 4;

          if (max1 <= this._inputSize) {
            chunk1 = this._input.substring(this._offset, max1);
          }

          if (chunk1 === 'true') {
            address2 = new TreeNode(this._input.substring(this._offset, this._offset + 4), this._offset, []);
            this._offset = this._offset + 4;
          } else {
            address2 = FAILURE;

            if (this._offset > this._failure) {
              this._failure = this._offset;
              this._expected = [];
            }

            if (this._offset === this._failure) {
              this._expected.push(['Maps::boolean_', '"true"']);
            }
          }

          if (address2 === FAILURE) {
            this._offset = index2;
            var chunk2 = null,
                max2 = this._offset + 5;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 === 'False') {
              address2 = new TreeNode(this._input.substring(this._offset, this._offset + 5), this._offset, []);
              this._offset = this._offset + 5;
            } else {
              address2 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::boolean_', '"False"']);
              }
            }

            if (address2 === FAILURE) {
              this._offset = index2;
              var chunk3 = null,
                  max3 = this._offset + 5;

              if (max3 <= this._inputSize) {
                chunk3 = this._input.substring(this._offset, max3);
              }

              if (chunk3 === 'false') {
                address2 = new TreeNode(this._input.substring(this._offset, this._offset + 5), this._offset, []);
                this._offset = this._offset + 5;
              } else {
                address2 = FAILURE;

                if (this._offset > this._failure) {
                  this._failure = this._offset;
                  this._expected = [];
                }

                if (this._offset === this._failure) {
                  this._expected.push(['Maps::boolean_', '"false"']);
                }
              }

              if (address2 === FAILURE) {
                this._offset = index2;
              }
            }
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read___();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_bool(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._boolean_[index0] = [address0, this._offset];
      return address0;
    },

    _read_null_() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._null_ = this._cache._null_ || {};
      var cached = this._cache._null_[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(3);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        var chunk0 = null,
            max0 = this._offset + 4;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === 'None') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 4), this._offset, []);
          this._offset = this._offset + 4;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::null_', '"None"']);
          }
        }

        if (address2 === FAILURE) {
          this._offset = index2;
          var chunk1 = null,
              max1 = this._offset + 3;

          if (max1 <= this._inputSize) {
            chunk1 = this._input.substring(this._offset, max1);
          }

          if (chunk1 === 'nil') {
            address2 = new TreeNode(this._input.substring(this._offset, this._offset + 3), this._offset, []);
            this._offset = this._offset + 3;
          } else {
            address2 = FAILURE;

            if (this._offset > this._failure) {
              this._failure = this._offset;
              this._expected = [];
            }

            if (this._offset === this._failure) {
              this._expected.push(['Maps::null_', '"nil"']);
            }
          }

          if (address2 === FAILURE) {
            this._offset = index2;
            var chunk2 = null,
                max2 = this._offset + 4;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 === 'null') {
              address2 = new TreeNode(this._input.substring(this._offset, this._offset + 4), this._offset, []);
              this._offset = this._offset + 4;
            } else {
              address2 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::null_', '"null"']);
              }
            }

            if (address2 === FAILURE) {
              this._offset = index2;
            }
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read___();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_null(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._null_[index0] = [address0, this._offset];
      return address0;
    },

    _read_nan_() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._nan_ = this._cache._nan_ || {};
      var cached = this._cache._nan_[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(3);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 3;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === 'NaN') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 3), this._offset, []);
          this._offset = this._offset + 3;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::nan_', '"NaN"']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read___();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_nan(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._nan_[index0] = [address0, this._offset];
      return address0;
    },

    _read_undefined_() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._undefined_ = this._cache._undefined_ || {};
      var cached = this._cache._undefined_[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(3);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var chunk0 = null,
            max0 = this._offset + 9;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === 'undefined') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 9), this._offset, []);
          this._offset = this._offset + 9;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::undefined_', '"undefined"']);
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read___();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_undefined(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._undefined_[index0] = [address0, this._offset];
      return address0;
    },

    _read_opt_() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache._opt_ = this._cache._opt_ || {};
      var cached = this._cache._opt_[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = new Array(3);
      var address1 = FAILURE;
      address1 = this._read___();

      if (address1 !== FAILURE) {
        elements0[0] = address1;
        var address2 = FAILURE;
        var index2 = this._offset;
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 === '+') {
          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address2 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::opt_', '"+"']);
          }
        }

        if (address2 === FAILURE) {
          this._offset = index2;
          var chunk1 = null,
              max1 = this._offset + 1;

          if (max1 <= this._inputSize) {
            chunk1 = this._input.substring(this._offset, max1);
          }

          if (chunk1 === '-') {
            address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
            this._offset = this._offset + 1;
          } else {
            address2 = FAILURE;

            if (this._offset > this._failure) {
              this._failure = this._offset;
              this._expected = [];
            }

            if (this._offset === this._failure) {
              this._expected.push(['Maps::opt_', '"-"']);
            }
          }

          if (address2 === FAILURE) {
            this._offset = index2;
            var chunk2 = null,
                max2 = this._offset + 1;

            if (max2 <= this._inputSize) {
              chunk2 = this._input.substring(this._offset, max2);
            }

            if (chunk2 === '*') {
              address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
              this._offset = this._offset + 1;
            } else {
              address2 = FAILURE;

              if (this._offset > this._failure) {
                this._failure = this._offset;
                this._expected = [];
              }

              if (this._offset === this._failure) {
                this._expected.push(['Maps::opt_', '"*"']);
              }
            }

            if (address2 === FAILURE) {
              this._offset = index2;
              var chunk3 = null,
                  max3 = this._offset + 1;

              if (max3 <= this._inputSize) {
                chunk3 = this._input.substring(this._offset, max3);
              }

              if (chunk3 === '/') {
                address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                this._offset = this._offset + 1;
              } else {
                address2 = FAILURE;

                if (this._offset > this._failure) {
                  this._failure = this._offset;
                  this._expected = [];
                }

                if (this._offset === this._failure) {
                  this._expected.push(['Maps::opt_', '"/"']);
                }
              }

              if (address2 === FAILURE) {
                this._offset = index2;
                var chunk4 = null,
                    max4 = this._offset + 2;

                if (max4 <= this._inputSize) {
                  chunk4 = this._input.substring(this._offset, max4);
                }

                if (chunk4 === '<=') {
                  address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                  this._offset = this._offset + 2;
                } else {
                  address2 = FAILURE;

                  if (this._offset > this._failure) {
                    this._failure = this._offset;
                    this._expected = [];
                  }

                  if (this._offset === this._failure) {
                    this._expected.push(['Maps::opt_', '"<="']);
                  }
                }

                if (address2 === FAILURE) {
                  this._offset = index2;
                  var chunk5 = null,
                      max5 = this._offset + 2;

                  if (max5 <= this._inputSize) {
                    chunk5 = this._input.substring(this._offset, max5);
                  }

                  if (chunk5 === '>=') {
                    address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                    this._offset = this._offset + 2;
                  } else {
                    address2 = FAILURE;

                    if (this._offset > this._failure) {
                      this._failure = this._offset;
                      this._expected = [];
                    }

                    if (this._offset === this._failure) {
                      this._expected.push(['Maps::opt_', '">="']);
                    }
                  }

                  if (address2 === FAILURE) {
                    this._offset = index2;
                    var chunk6 = null,
                        max6 = this._offset + 2;

                    if (max6 <= this._inputSize) {
                      chunk6 = this._input.substring(this._offset, max6);
                    }

                    if (chunk6 === '!=') {
                      address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                      this._offset = this._offset + 2;
                    } else {
                      address2 = FAILURE;

                      if (this._offset > this._failure) {
                        this._failure = this._offset;
                        this._expected = [];
                      }

                      if (this._offset === this._failure) {
                        this._expected.push(['Maps::opt_', '"!="']);
                      }
                    }

                    if (address2 === FAILURE) {
                      this._offset = index2;
                      var chunk7 = null,
                          max7 = this._offset + 2;

                      if (max7 <= this._inputSize) {
                        chunk7 = this._input.substring(this._offset, max7);
                      }

                      if (chunk7 === '==') {
                        address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                        this._offset = this._offset + 2;
                      } else {
                        address2 = FAILURE;

                        if (this._offset > this._failure) {
                          this._failure = this._offset;
                          this._expected = [];
                        }

                        if (this._offset === this._failure) {
                          this._expected.push(['Maps::opt_', '"=="']);
                        }
                      }

                      if (address2 === FAILURE) {
                        this._offset = index2;
                        var chunk8 = null,
                            max8 = this._offset + 1;

                        if (max8 <= this._inputSize) {
                          chunk8 = this._input.substring(this._offset, max8);
                        }

                        if (chunk8 === '=') {
                          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                          this._offset = this._offset + 1;
                        } else {
                          address2 = FAILURE;

                          if (this._offset > this._failure) {
                            this._failure = this._offset;
                            this._expected = [];
                          }

                          if (this._offset === this._failure) {
                            this._expected.push(['Maps::opt_', '"="']);
                          }
                        }

                        if (address2 === FAILURE) {
                          this._offset = index2;
                          var chunk9 = null,
                              max9 = this._offset + 1;

                          if (max9 <= this._inputSize) {
                            chunk9 = this._input.substring(this._offset, max9);
                          }

                          if (chunk9 === '>') {
                            address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                            this._offset = this._offset + 1;
                          } else {
                            address2 = FAILURE;

                            if (this._offset > this._failure) {
                              this._failure = this._offset;
                              this._expected = [];
                            }

                            if (this._offset === this._failure) {
                              this._expected.push(['Maps::opt_', '">"']);
                            }
                          }

                          if (address2 === FAILURE) {
                            this._offset = index2;
                            var chunk10 = null,
                                max10 = this._offset + 1;

                            if (max10 <= this._inputSize) {
                              chunk10 = this._input.substring(this._offset, max10);
                            }

                            if (chunk10 === '<') {
                              address2 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
                              this._offset = this._offset + 1;
                            } else {
                              address2 = FAILURE;

                              if (this._offset > this._failure) {
                                this._failure = this._offset;
                                this._expected = [];
                              }

                              if (this._offset === this._failure) {
                                this._expected.push(['Maps::opt_', '"<"']);
                              }
                            }

                            if (address2 === FAILURE) {
                              this._offset = index2;
                              var chunk11 = null,
                                  max11 = this._offset + 2;

                              if (max11 <= this._inputSize) {
                                chunk11 = this._input.substring(this._offset, max11);
                              }

                              if (chunk11 === 'LT') {
                                address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                this._offset = this._offset + 2;
                              } else {
                                address2 = FAILURE;

                                if (this._offset > this._failure) {
                                  this._failure = this._offset;
                                  this._expected = [];
                                }

                                if (this._offset === this._failure) {
                                  this._expected.push(['Maps::opt_', '"LT"']);
                                }
                              }

                              if (address2 === FAILURE) {
                                this._offset = index2;
                                var chunk12 = null,
                                    max12 = this._offset + 2;

                                if (max12 <= this._inputSize) {
                                  chunk12 = this._input.substring(this._offset, max12);
                                }

                                if (chunk12 === 'GT') {
                                  address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                  this._offset = this._offset + 2;
                                } else {
                                  address2 = FAILURE;

                                  if (this._offset > this._failure) {
                                    this._failure = this._offset;
                                    this._expected = [];
                                  }

                                  if (this._offset === this._failure) {
                                    this._expected.push(['Maps::opt_', '"GT"']);
                                  }
                                }

                                if (address2 === FAILURE) {
                                  this._offset = index2;
                                  var chunk13 = null,
                                      max13 = this._offset + 2;

                                  if (max13 <= this._inputSize) {
                                    chunk13 = this._input.substring(this._offset, max13);
                                  }

                                  if (chunk13 === 'LE') {
                                    address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                    this._offset = this._offset + 2;
                                  } else {
                                    address2 = FAILURE;

                                    if (this._offset > this._failure) {
                                      this._failure = this._offset;
                                      this._expected = [];
                                    }

                                    if (this._offset === this._failure) {
                                      this._expected.push(['Maps::opt_', '"LE"']);
                                    }
                                  }

                                  if (address2 === FAILURE) {
                                    this._offset = index2;
                                    var chunk14 = null,
                                        max14 = this._offset + 2;

                                    if (max14 <= this._inputSize) {
                                      chunk14 = this._input.substring(this._offset, max14);
                                    }

                                    if (chunk14 === 'GE') {
                                      address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                      this._offset = this._offset + 2;
                                    } else {
                                      address2 = FAILURE;

                                      if (this._offset > this._failure) {
                                        this._failure = this._offset;
                                        this._expected = [];
                                      }

                                      if (this._offset === this._failure) {
                                        this._expected.push(['Maps::opt_', '"GE"']);
                                      }
                                    }

                                    if (address2 === FAILURE) {
                                      this._offset = index2;
                                      var chunk15 = null,
                                          max15 = this._offset + 2;

                                      if (max15 <= this._inputSize) {
                                        chunk15 = this._input.substring(this._offset, max15);
                                      }

                                      if (chunk15 === 'EQ') {
                                        address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                        this._offset = this._offset + 2;
                                      } else {
                                        address2 = FAILURE;

                                        if (this._offset > this._failure) {
                                          this._failure = this._offset;
                                          this._expected = [];
                                        }

                                        if (this._offset === this._failure) {
                                          this._expected.push(['Maps::opt_', '"EQ"']);
                                        }
                                      }

                                      if (address2 === FAILURE) {
                                        this._offset = index2;
                                        var chunk16 = null,
                                            max16 = this._offset + 2;

                                        if (max16 <= this._inputSize) {
                                          chunk16 = this._input.substring(this._offset, max16);
                                        }

                                        if (chunk16 === 'NE') {
                                          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                          this._offset = this._offset + 2;
                                        } else {
                                          address2 = FAILURE;

                                          if (this._offset > this._failure) {
                                            this._failure = this._offset;
                                            this._expected = [];
                                          }

                                          if (this._offset === this._failure) {
                                            this._expected.push(['Maps::opt_', '"NE"']);
                                          }
                                        }

                                        if (address2 === FAILURE) {
                                          this._offset = index2;
                                          var chunk17 = null,
                                              max17 = this._offset + 2;

                                          if (max17 <= this._inputSize) {
                                            chunk17 = this._input.substring(this._offset, max17);
                                          }

                                          if (chunk17 === 'lt') {
                                            address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                            this._offset = this._offset + 2;
                                          } else {
                                            address2 = FAILURE;

                                            if (this._offset > this._failure) {
                                              this._failure = this._offset;
                                              this._expected = [];
                                            }

                                            if (this._offset === this._failure) {
                                              this._expected.push(['Maps::opt_', '"lt"']);
                                            }
                                          }

                                          if (address2 === FAILURE) {
                                            this._offset = index2;
                                            var chunk18 = null,
                                                max18 = this._offset + 2;

                                            if (max18 <= this._inputSize) {
                                              chunk18 = this._input.substring(this._offset, max18);
                                            }

                                            if (chunk18 === 'gt') {
                                              address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                              this._offset = this._offset + 2;
                                            } else {
                                              address2 = FAILURE;

                                              if (this._offset > this._failure) {
                                                this._failure = this._offset;
                                                this._expected = [];
                                              }

                                              if (this._offset === this._failure) {
                                                this._expected.push(['Maps::opt_', '"gt"']);
                                              }
                                            }

                                            if (address2 === FAILURE) {
                                              this._offset = index2;
                                              var chunk19 = null,
                                                  max19 = this._offset + 2;

                                              if (max19 <= this._inputSize) {
                                                chunk19 = this._input.substring(this._offset, max19);
                                              }

                                              if (chunk19 === 'le') {
                                                address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                this._offset = this._offset + 2;
                                              } else {
                                                address2 = FAILURE;

                                                if (this._offset > this._failure) {
                                                  this._failure = this._offset;
                                                  this._expected = [];
                                                }

                                                if (this._offset === this._failure) {
                                                  this._expected.push(['Maps::opt_', '"le"']);
                                                }
                                              }

                                              if (address2 === FAILURE) {
                                                this._offset = index2;
                                                var chunk20 = null,
                                                    max20 = this._offset + 2;

                                                if (max20 <= this._inputSize) {
                                                  chunk20 = this._input.substring(this._offset, max20);
                                                }

                                                if (chunk20 === 'ge') {
                                                  address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                  this._offset = this._offset + 2;
                                                } else {
                                                  address2 = FAILURE;

                                                  if (this._offset > this._failure) {
                                                    this._failure = this._offset;
                                                    this._expected = [];
                                                  }

                                                  if (this._offset === this._failure) {
                                                    this._expected.push(['Maps::opt_', '"ge"']);
                                                  }
                                                }

                                                if (address2 === FAILURE) {
                                                  this._offset = index2;
                                                  var chunk21 = null,
                                                      max21 = this._offset + 2;

                                                  if (max21 <= this._inputSize) {
                                                    chunk21 = this._input.substring(this._offset, max21);
                                                  }

                                                  if (chunk21 === 'eq') {
                                                    address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                    this._offset = this._offset + 2;
                                                  } else {
                                                    address2 = FAILURE;

                                                    if (this._offset > this._failure) {
                                                      this._failure = this._offset;
                                                      this._expected = [];
                                                    }

                                                    if (this._offset === this._failure) {
                                                      this._expected.push(['Maps::opt_', '"eq"']);
                                                    }
                                                  }

                                                  if (address2 === FAILURE) {
                                                    this._offset = index2;
                                                    var chunk22 = null,
                                                        max22 = this._offset + 2;

                                                    if (max22 <= this._inputSize) {
                                                      chunk22 = this._input.substring(this._offset, max22);
                                                    }

                                                    if (chunk22 === 'ne') {
                                                      address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                      this._offset = this._offset + 2;
                                                    } else {
                                                      address2 = FAILURE;

                                                      if (this._offset > this._failure) {
                                                        this._failure = this._offset;
                                                        this._expected = [];
                                                      }

                                                      if (this._offset === this._failure) {
                                                        this._expected.push(['Maps::opt_', '"ne"']);
                                                      }
                                                    }

                                                    if (address2 === FAILURE) {
                                                      this._offset = index2;
                                                      var chunk23 = null,
                                                          max23 = this._offset + 2;

                                                      if (max23 <= this._inputSize) {
                                                        chunk23 = this._input.substring(this._offset, max23);
                                                      }

                                                      if (chunk23 === 'in') {
                                                        address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                        this._offset = this._offset + 2;
                                                      } else {
                                                        address2 = FAILURE;

                                                        if (this._offset > this._failure) {
                                                          this._failure = this._offset;
                                                          this._expected = [];
                                                        }

                                                        if (this._offset === this._failure) {
                                                          this._expected.push(['Maps::opt_', '"in"']);
                                                        }
                                                      }

                                                      if (address2 === FAILURE) {
                                                        this._offset = index2;
                                                        var chunk24 = null,
                                                            max24 = this._offset + 2;

                                                        if (max24 <= this._inputSize) {
                                                          chunk24 = this._input.substring(this._offset, max24);
                                                        }

                                                        if (chunk24 === 'IN') {
                                                          address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                          this._offset = this._offset + 2;
                                                        } else {
                                                          address2 = FAILURE;

                                                          if (this._offset > this._failure) {
                                                            this._failure = this._offset;
                                                            this._expected = [];
                                                          }

                                                          if (this._offset === this._failure) {
                                                            this._expected.push(['Maps::opt_', '"IN"']);
                                                          }
                                                        }

                                                        if (address2 === FAILURE) {
                                                          this._offset = index2;
                                                          var chunk25 = null,
                                                              max25 = this._offset + 2;

                                                          if (max25 <= this._inputSize) {
                                                            chunk25 = this._input.substring(this._offset, max25);
                                                          }

                                                          if (chunk25 === 'or') {
                                                            address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                            this._offset = this._offset + 2;
                                                          } else {
                                                            address2 = FAILURE;

                                                            if (this._offset > this._failure) {
                                                              this._failure = this._offset;
                                                              this._expected = [];
                                                            }

                                                            if (this._offset === this._failure) {
                                                              this._expected.push(['Maps::opt_', '"or"']);
                                                            }
                                                          }

                                                          if (address2 === FAILURE) {
                                                            this._offset = index2;
                                                            var chunk26 = null,
                                                                max26 = this._offset + 2;

                                                            if (max26 <= this._inputSize) {
                                                              chunk26 = this._input.substring(this._offset, max26);
                                                            }

                                                            if (chunk26 === 'OR') {
                                                              address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                              this._offset = this._offset + 2;
                                                            } else {
                                                              address2 = FAILURE;

                                                              if (this._offset > this._failure) {
                                                                this._failure = this._offset;
                                                                this._expected = [];
                                                              }

                                                              if (this._offset === this._failure) {
                                                                this._expected.push(['Maps::opt_', '"OR"']);
                                                              }
                                                            }

                                                            if (address2 === FAILURE) {
                                                              this._offset = index2;
                                                              var chunk27 = null,
                                                                  max27 = this._offset + 2;

                                                              if (max27 <= this._inputSize) {
                                                                chunk27 = this._input.substring(this._offset, max27);
                                                              }

                                                              if (chunk27 === '||') {
                                                                address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                                this._offset = this._offset + 2;
                                                              } else {
                                                                address2 = FAILURE;

                                                                if (this._offset > this._failure) {
                                                                  this._failure = this._offset;
                                                                  this._expected = [];
                                                                }

                                                                if (this._offset === this._failure) {
                                                                  this._expected.push(['Maps::opt_', '"||"']);
                                                                }
                                                              }

                                                              if (address2 === FAILURE) {
                                                                this._offset = index2;
                                                                var chunk28 = null,
                                                                    max28 = this._offset + 3;

                                                                if (max28 <= this._inputSize) {
                                                                  chunk28 = this._input.substring(this._offset, max28);
                                                                }

                                                                if (chunk28 === 'and') {
                                                                  address2 = new TreeNode(this._input.substring(this._offset, this._offset + 3), this._offset, []);
                                                                  this._offset = this._offset + 3;
                                                                } else {
                                                                  address2 = FAILURE;

                                                                  if (this._offset > this._failure) {
                                                                    this._failure = this._offset;
                                                                    this._expected = [];
                                                                  }

                                                                  if (this._offset === this._failure) {
                                                                    this._expected.push(['Maps::opt_', '"and"']);
                                                                  }
                                                                }

                                                                if (address2 === FAILURE) {
                                                                  this._offset = index2;
                                                                  var chunk29 = null,
                                                                      max29 = this._offset + 3;

                                                                  if (max29 <= this._inputSize) {
                                                                    chunk29 = this._input.substring(this._offset, max29);
                                                                  }

                                                                  if (chunk29 === 'AND') {
                                                                    address2 = new TreeNode(this._input.substring(this._offset, this._offset + 3), this._offset, []);
                                                                    this._offset = this._offset + 3;
                                                                  } else {
                                                                    address2 = FAILURE;

                                                                    if (this._offset > this._failure) {
                                                                      this._failure = this._offset;
                                                                      this._expected = [];
                                                                    }

                                                                    if (this._offset === this._failure) {
                                                                      this._expected.push(['Maps::opt_', '"AND"']);
                                                                    }
                                                                  }

                                                                  if (address2 === FAILURE) {
                                                                    this._offset = index2;
                                                                    var chunk30 = null,
                                                                        max30 = this._offset + 2;

                                                                    if (max30 <= this._inputSize) {
                                                                      chunk30 = this._input.substring(this._offset, max30);
                                                                    }

                                                                    if (chunk30 === '&&') {
                                                                      address2 = new TreeNode(this._input.substring(this._offset, this._offset + 2), this._offset, []);
                                                                      this._offset = this._offset + 2;
                                                                    } else {
                                                                      address2 = FAILURE;

                                                                      if (this._offset > this._failure) {
                                                                        this._failure = this._offset;
                                                                        this._expected = [];
                                                                      }

                                                                      if (this._offset === this._failure) {
                                                                        this._expected.push(['Maps::opt_', '"&&"']);
                                                                      }
                                                                    }

                                                                    if (address2 === FAILURE) {
                                                                      this._offset = index2;
                                                                    }
                                                                  }
                                                                }
                                                              }
                                                            }
                                                          }
                                                        }
                                                      }
                                                    }
                                                  }
                                                }
                                              }
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }

        if (address2 !== FAILURE) {
          elements0[1] = address2;
          var address3 = FAILURE;
          address3 = this._read___();

          if (address3 !== FAILURE) {
            elements0[2] = address3;
          } else {
            elements0 = null;
            this._offset = index1;
          }
        } else {
          elements0 = null;
          this._offset = index1;
        }
      } else {
        elements0 = null;
        this._offset = index1;
      }

      if (elements0 === null) {
        address0 = FAILURE;
      } else {
        address0 = this._actions.make_opt(this._input, index1, this._offset, elements0);
        this._offset = this._offset;
      }

      this._cache._opt_[index0] = [address0, this._offset];
      return address0;
    },

    _read___() {
      var address0 = FAILURE,
          index0 = this._offset;
      this._cache.___ = this._cache.___ || {};
      var cached = this._cache.___[index0];

      if (cached) {
        this._offset = cached[1];
        return cached[0];
      }

      var index1 = this._offset,
          elements0 = [],
          address1 = null;

      while (true) {
        var chunk0 = null,
            max0 = this._offset + 1;

        if (max0 <= this._inputSize) {
          chunk0 = this._input.substring(this._offset, max0);
        }

        if (chunk0 !== null && /^[\s]/.test(chunk0)) {
          address1 = new TreeNode(this._input.substring(this._offset, this._offset + 1), this._offset, []);
          this._offset = this._offset + 1;
        } else {
          address1 = FAILURE;

          if (this._offset > this._failure) {
            this._failure = this._offset;
            this._expected = [];
          }

          if (this._offset === this._failure) {
            this._expected.push(['Maps::__', '[\\s]']);
          }
        }

        if (address1 !== FAILURE) {
          elements0.push(address1);
        } else {
          break;
        }
      }

      if (elements0.length >= 0) {
        address0 = new TreeNode(this._input.substring(index1, this._offset), index1, elements0);
        this._offset = this._offset;
      } else {
        address0 = FAILURE;
      }

      this._cache.___[index0] = [address0, this._offset];
      return address0;
    }

  };

  var Parser = function (input, actions, types) {
    this._input = input;
    this._inputSize = input.length;
    this._actions = actions;
    this._types = types;
    this._offset = 0;
    this._cache = {};
    this._failure = 0;
    this._expected = [];
  };

  Parser.prototype.parse = function () {
    var tree = this._read_comp_expression();

    if (tree !== FAILURE && this._offset === this._inputSize) {
      return tree;
    }

    if (this._expected.length === 0) {
      this._failure = this._offset;

      this._expected.push(['Maps', '<EOF>']);
    }

    this.constructor.lastError = {
      offset: this._offset,
      expected: this._expected
    };
    throw new SyntaxError(formatError(this._input, this._failure, this._expected));
  };

  Object.assign(Parser.prototype, Grammar);

  function parse(input, options) {
    options = options || {};
    var parser = new Parser(input, options.actions, options.types);
    return parser.parse();
  }

  function formatError(input, offset, expected) {
    var lines = input.split(/\n/g),
        lineNo = 0,
        position = 0;

    while (position <= offset) {
      position += lines[lineNo].length + 1;
      lineNo += 1;
    }

    var line = lines[lineNo - 1],
        message = 'Line ' + lineNo + ': expected one of:\n\n';

    for (var i = 0; i < expected.length; i++) {
      message += '    - ' + expected[i][1] + ' from ' + expected[i][0] + '\n';
    }

    var number = lineNo.toString();

    while (number.length < 6) number = ' ' + number;

    message += '\n' + number + ' | ' + line + '\n';
    position -= line.length + 10;

    while (position < offset) {
      message += ' ';
      position += 1;
    }

    return message + '^';
  }

  function inherit(subclass, parent) {
    function chain() {}

    ;
    chain.prototype = parent.prototype;
    subclass.prototype = new chain();
    subclass.prototype.constructor = subclass;
  }

  var exported = {
    Grammar: Grammar,
    Parser: Parser,
    parse: parse
  };

  if (typeof require === 'function' && typeof exports === 'object') {
    Object.assign(exports, exported);
  } else {
    var ns = typeof this === 'undefined' ? window : this;
    ns.Maps = exported;
  }
})();
